internal/pkg/call/call_test.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package call
+
+import (
+	"reflect"
+	"testing"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/analysistest"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+	"golang.org/x/tools/go/ssa"
+)
+
+func TestRegularCallReferredBy(t *testing.T) {
+	cases := []struct {
+		desc    string
+		pattern string
+		r       Referrer
+		want    bool
+	}{
+		{
+			"call without args",
+			"noarg",
+			fakeReferrer{true},
+			false,
+		},
+		{
+			"call with one arg that is referred to by a referrer",
+			"onearg",
+			fakeReferrer{true},
+			true,
+		},
+		{
+			"call with one arg that isn't referred to",
+			"onearg",
+			fakeReferrer{false},
+			false,
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.desc, func(t *testing.T) {
+			dir := analysistest.TestData()
+			r := analysistest.Run(t, dir, testAnalyzer, tt.pattern)
+			call := Regular(getCall(t, r))
+			got := call.ReferredBy(tt.r)
+			if got != tt.want {
+				t.Errorf("call.ReferredBy(%v) == %v, want %v", call, got, tt.want)
+			}
+		})
+	}
+}
+
+func getCall(t *testing.T, r []*analysistest.Result) *ssa.Call {
+	t.Helper()
+	if len(r) != 1 {
+		t.Fatalf("Got %d results, want one", len(r))
+	}
+	a, ok := r[0].Result.(analyzerResult)
+	if !ok {
+		t.Fatalf("Got result of type %T, want analyzerResult", a)
+	}
+	if len(a.calls) != 1 {
+		t.Fatalf("Got %d calls, want one", len(a.calls))
+	}
+	return a.calls[0]
+}
+
+type fakeReferrer struct {
+	hasPathTo bool
+}
+
+func (f fakeReferrer) RefersTo(node ssa.Node) bool { return f.hasPathTo }
+
+var testAnalyzer = &analysis.Analyzer{
+	Name:       "call",
+	Run:        run,
+	Doc:        "test harness for call",
+	Requires:   []*analysis.Analyzer{buildssa.Analyzer},
+	ResultType: reflect.TypeOf(analyzerResult{}),
+}
+
+type analyzerResult struct {
+	calls []*ssa.Call
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	in := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+	var result analyzerResult
+	for _, fn := range in.SrcFuncs {
+		for _, b := range fn.Blocks {
+			for _, i := range b.Instrs {
+				switch v := i.(type) {
+				case *ssa.Call:
+					result.calls = append(result.calls, v)
```

Comment:
Here, instead of collecting results, do `v.String()` - this will emit a finding (something like: "Referred by X"). Assuming that you have a corresponding `want` comment and the source code input has referrer X, this will be all that we need to test for.

This will allow us to remove the fake referrer, but rather rely on the source code input to setup the testcase.

Reply:
I'm not sure how to set this up. I can't just do `v.String()`, I would need to do `pass.Reportf(...)`. I'm not sure what I would put in the "want" comment though... more importantly, I don't know how I would produce a report that would allow me to test that `RegularCall.ReferredBy` is working properly. I need to create a `RegularCall` at some point, which I could do in the `Analyzer` we have here. But then, I would still need the fakeReferrer... I'm also not sure how I would cover the case where the referrer *does not* refer to the call.